gitdiff/patch_header.go

```diff
@@ -369,3 +373,61 @@ func parseHeaderMail(mailLine string, r io.Reader) (*PatchHeader, error) {
 
 	return h, nil
 }
+
+func cleanupSubject(s string) string {
+	r := []rune(s)
+
+	// This is meant to be compatible with
+	// https://github.com/git/git/blob/master/mailinfo.c:cleanup_subject().
+	// If compatibility with `git am` drifts, go there to see if there
+	// are any updates.
+
+	var at int
+	for at = 0; at < len(r); {
+		switch r[at] {
+		case 'r', 'R':
+			// Detect re:, Re:, rE: and RE:
+			if at+3 > len(r) {
+				break
+			}
+			if (r[at+1] == 'e' || r[at+1] == 'E') &&
+				r[at+2] == ':' {
+				at += 3
+				continue
+			}
+
+		case ' ', '\t', ':':
+			// Delete whitespace and duplicate ':' characters
+			at++
+			continue
+
+		case '[':
+			// Look for closing parenthesis
+			var closepos int
+			for j := at + 1; j < len(r); j++ {
+				if r[j] == ']' {
+					closepos = j
+					break
+				}
+			}
+
+			if closepos == 0 {
+				break
+			}
+
+			at = closepos + 1
+			continue
```

Comment:
style: I think this could be simplified:

```suggestion
			j := at + 1
			for j < len(r); j++ {
				if r[j] == ']' {
					break
				}
			}
			if j < len(r) {
				at = j + 1
				continue
			}
```